stdlib/public/core/StringStorage.swift

```diff
@@ -16,59 +16,103 @@ import SwiftShims
 @usableFromInline
 internal class _AbstractStringStorage: __SwiftNativeNSString, _NSStringCore {
   // Abstract interface
-  internal var asString: String { get { Builtin.unreachable() } }
-  internal var count: Int { get { Builtin.unreachable() } }
-  internal func getOrComputeBreadcrumbs() -> _StringBreadcrumbs {
-    Builtin.unreachable()
-  }
+  internal var asGuts: _StringGuts { @_effects(readonly) get { Builtin.unreachable() } }
+  final internal var asString: String { @_effects(readonly) get { return String(asGuts) } }
+  internal var count: Int { @_effects(readonly) get { Builtin.unreachable() } }
```

Comment:
Dynamically dispatching this allows us to statically dispatch the asString/asGuts calls inside the implementations, which allows us to avoid some refcounting